### DISCUSSION TITLE
Taichi user experience lifting wish list

### DISCUSSION THREAD

**Discussion by neozhaoliang** [taichi-dev/taichi#5446] [General]
# Taichi wish list

# 1

+  [ ] Support syntax like `ti.f32(x)` for variables.

**Comment by neozhaoliang** [comment]
# 2

+ [ ] Call a `ti.func` in Python scope.

**Comment by neozhaoliang** [comment]
# 3

+ [ ] Call a kernel function inside another kernel.
    
    For example:
    ```python
    @ti.kernel
    def render():
        grid.fill(0)  # clear a field
     ```

**Comment by neozhaoliang** [comment]
# 4

+ [ ] Create / delete local fields in a kernel (top level)

    For example:
```python
@ti.kernel
def task():
    field = ti.field(int,shape=(10, 10))
   
    for i in range(10):
        field[i, i] = 1
```

**Comment by neozhaoliang** [comment]
# 5

+  [ ] Create a field from `numpy.ndarry` or `torch.tensor`.
    
    For example:
    ```python
    arr = np.random.random((100, 100))
    field = ti.field.from_nparray(arr)
    ```

**Comment by neozhaoliang** [comment]
# 6

+  [ ] No need to explicitly write down the `(name, val)` pair when initializing a `struct`:
    
    For example:
    ```python
    import taichi as ti
    ti.init()

    vec2_t = ti.types.vector(2, float)

    @ti.dataclass
    class Foo:
        pos: vec2_t
        val: float
        delay: float

    @ti.kernel
    def test():
        F = Foo((0.0, 0.0), 1.0)
        print(F.pos, F.val, F.delay)

    test()
    ```

**Comment by neozhaoliang** [comment]
# 7

+ [ ] Raw kernel, like NV Warp/CUDA kernel, proposed by @yuanming-hu

**Comment by neozhaoliang** [comment]
# 8

+ [ ] Global time counter like shadertoy's `iTime`, usually used for animations.
   
    For example:

    ```python
    tick = ti.Clock()  # start at 0
    tick.pause()
    tick.continue()   # continue from last pause
    tick.stop()
    ```

**Reply by yuanming-hu** [reply]
Could you add a usage example here?

**Reply by bobcao3** [reply]
Do not like this idea... Why can't the user just use Python builtin clocks

**Reply by neozhaoliang** [reply]
When an animations is paused or resumed by user input, we hope the time counter resumes from the last time it's paused. I don't know if the built-in  module support this? Also this counter should be different if there are several animations in one program?

**Comment by neozhaoliang** [comment]
# 9

+ [ ] Display  line number ranges when there are multiply offloaded tasks in a `KernelProfiler`.

**Comment by neozhaoliang** [comment]
# 10

+ [ ] Assign name to `ti.field`, so that when debugging the out of indexing error it can print the field name, not just  `SNode_xxx`.

**Reply by neozhaoliang** [reply]
open an issue

**Comment by neozhaoliang** [comment]
# 11

+ [ ] Make `examples` a top level directory.

(and as well as `docs`, `tests` folders?)
